src/dns.rs

```diff
@@ -99,46 +149,37 @@ impl Resolver {
         (resolver, background)
     }
 
-    pub fn resolve_one_ip(&self, host: &Addr) -> IpAddrFuture {
-        match host {
-            Addr::Name(n) => {
-                let name = n.name();
-                let ctx = ResolveOneCtx(name.clone());
-                let f = ::logging::context_future(ctx, self.lookup_ip(name));
-                IpAddrFuture::DNS(Box::new(f))
-            }
-            Addr::Socket(addr) => IpAddrFuture::Fixed(addr.ip()),
-        }
-    }
+    pub fn resolve_all_ips(&self, deadline: Instant, name: &Name) -> IpAddrListFuture {
+        let lookup = self.resolver.lookup_ip(name.as_ref());
 
-    pub fn resolve_all_ips(&self, deadline: Instant, host: &Name) -> IpAddrListFuture {
-        let name = host.clone();
-        let lookup = self.lookup_ip(&name);
+        // FIXME this delay logic is really confusing...
         let f = Delay::new(deadline)
             .then(move |_| {
                 trace!("after delay");
                 lookup
             })
             .then(move |result| {
                 trace!("completed with {:?}", &result);
-                match result {
-                    Ok(ips) => Ok(Response::Exists(ips)),
-                    Err(e) => {
-                        if let &ResolveErrorKind::NoRecordsFound { valid_until, .. } = e.kind() {
-                            Ok(Response::DoesNotExist { retry_after: valid_until })
-                        } else {
-                            Err(e)
-                        }
+                result.map(Response::Exists).or_else(|e| {
+                    if let &ResolveErrorKind::NoRecordsFound { valid_until, .. } = e.kind() {
+                        Ok(Response::DoesNotExist { retry_after: valid_until })
+                    } else {
+                        Err(e)
                     }
-                }
+                })
             });
-        Box::new(::logging::context_future(ResolveAllCtx(name), f))
+
+        Box::new(::logging::context_future(Ctx(name.clone()), f))
     }
 
-    fn lookup_ip(&self, name: &Name)
-        -> impl Future<Item = LookupIp, Error = ResolveError>
-    {
-        self.resolver.lookup_ip(name.as_ref())
+    pub fn resolve_one_ip(&self, name: &Name) -> IpAddrFuture {
+        let f = self.resolver.lookup_ip(name.as_ref());
+        IpAddrFuture(::logging::context_future(Ctx(name.clone()), f))
+    }
+
+    pub fn refine(&self, name: &Name) -> RefineFuture {
```

Comment:
Personally, I'd like to see a comment either here or on `RefineFuture` --- it took me a while to work out what this was actually doing.